app/scripts/controllers/permissions/background-api.js

```diff
@@ -1,45 +1,131 @@
 import nanoid from 'nanoid';
+import { MethodNames } from '@metamask/permission-controller';
 import {
-  CaveatTypes,
-  RestrictedMethods,
-} from '../../../../shared/constants/permissions';
-import { CaveatFactories, PermissionNames } from './specifications';
+  Caip25CaveatType,
+  Caip25EndowmentPermissionName,
+  getEthAccounts,
+  setEthAccounts,
+  getPermittedEthChainIds,
+  setPermittedEthChainIds,
+} from '@metamask/multichain';
+import { RestrictedMethods } from '../../../../shared/constants/permissions';
+import { PermissionNames } from './specifications';
 
-export function getPermissionBackgroundApiMethods(permissionController) {
+const snapsPrefixes = ['npm:', 'local:'];
+const isSnap = (origin) =>
+  snapsPrefixes.some((prefix) => origin.startsWith(prefix));
+
+export function getPermissionBackgroundApiMethods({
+  permissionController,
+  approvalController,
+}) {
+  // To add more than one account when already connected to the dapp
   const addMoreAccounts = (origin, accounts) => {
-    const caveat = CaveatFactories.restrictReturnedAccounts(accounts);
-
-    permissionController.grantPermissionsIncremental({
-      subject: { origin },
-      approvedPermissions: {
-        [RestrictedMethods.eth_accounts]: { caveats: [caveat] },
-      },
-    });
+    let caip25Caveat;
+    try {
+      caip25Caveat = permissionController.getCaveat(
+        origin,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    } catch (err) {
+      // noop
+    }
+
+    if (!caip25Caveat) {
+      throw new Error(
+        `Cannot add account permissions for origin "${origin}": no permission currently exists for this origin.`,
+      );
+    }
+
+    const ethAccounts = getEthAccounts(caip25Caveat.value);
+
+    const updatedEthAccounts = Array.from(
+      new Set([...ethAccounts, ...accounts]),
+    );
+
+    const updatedCaveatValue = setEthAccounts(
+      caip25Caveat.value,
+      updatedEthAccounts,
+    );
+
+    permissionController.updateCaveat(
+      origin,
+      Caip25EndowmentPermissionName,
+      Caip25CaveatType,
+      updatedCaveatValue,
+    );
   };
 
   const addMoreChains = (origin, chainIds) => {
-    const caveat = CaveatFactories.restrictNetworkSwitching(chainIds);
-
-    permissionController.grantPermissionsIncremental({
-      subject: { origin },
-      approvedPermissions: {
-        [PermissionNames.permittedChains]: { caveats: [caveat] },
-      },
-    });
+    let caip25Caveat;
+    try {
+      caip25Caveat = permissionController.getCaveat(
+        origin,
+        Caip25EndowmentPermissionName,
+        Caip25CaveatType,
+      );
+    } catch (err) {
+      // noop
+    }
+
+    if (!caip25Caveat) {
+      throw new Error(
+        `Cannot add chain permissions for origin "${origin}": no permission currently exists for this origin.`,
+      );
+    }
+
+    // get the list of permitted eth accounts before we modify the permitted chains and potentially lose some
```

Comment:
Why would we lose accounts when modifying permitted chains?

Reply:
If the new set of chains do not overlap at all with the old set of chains, then the accounts stored in the old scopes will get lost when `setPermittedEthChainIds` is called. Because of that, we need to grab the accounts before we modify the scopes and then restore them

Reply:
updated the inline comment here https://github.com/MetaMask/metamask-extension/pull/27847/commits/fb118fe7420fb53bf41c985bd4a2a82e5e96c921

Reply:
But this method is called `addMoreChains`, there is no way for this to take away access to chains. Unless I'm misreading something.